json/src/pointer.rs

```diff
@@ -0,0 +1,141 @@
+
+    /// Looks up a value by a JSON Pointer and returns a mutable reference to
+    /// that value.
+    ///
+    /// JSON Pointer defines a string syntax for identifying a specific value
+    /// within a JavaScript Object Notation (JSON) document.
+    ///
+    /// A Pointer is a Unicode string with the reference tokens separated by `/`.
+    /// Inside tokens `/` is replaced by `~1` and `~` is replaced by `~0`. The
+    /// addressed value is returned and if there is no such value `None` is
+    /// returned.
+    fn pointer_mut<'a>(&'a mut self, pointer: &str) -> Option<&'a mut Value>;
+
+    /// Looks up a value by a JSON Pointer while consuming the object to return
+    /// the value as owned, immutable data.
+    ///
+    /// JSON Pointer defines a string syntax for identifying a specific value
+    /// within a JavaScript Object Notation (JSON) document.
+    ///
+    /// A Pointer is a Unicode string with the reference tokens separated by `/`.
+    /// Inside tokens `/` is replaced by `~1` and `~` is replaced by `~0`. The
+    /// addressed value is returned and if there is no such value `None` is
+    /// returned.
+    fn pointer_owned(self, pointer: &str) -> Option<Value>;
+}
+
+fn parse_index(s: &str) -> Option<usize> {
+    if s.starts_with('+') || (s.starts_with('0') && s.len() != 1) {
+        return None;
+    }
+    s.parse().ok()
+}
+
+impl Pointer for Value {
+    fn pointer<'a>(&'a self, pointer: &str) -> Option<&'a Value> {
+        if pointer == "" {
+            return Some(self);
+        }
+        if !pointer.starts_with('/') {
+            return None;
+        }
+        let tokens = pointer.split('/').skip(1).map(|x| x.replace("~1", "/").replace("~0", "~"));
+        let mut target = self;
+
+        for token in tokens {
+            let target_opt = match *target {
+                Value::Object(ref map) => map.get(&token[..]),
+                Value::Array(ref list) => {
+                    parse_index(&token[..]).and_then(|x| list.get(x))
+                }
+                _ => return None,
+            };
+            if let Some(t) = target_opt {
+                target = t;
+            } else {
+                return None;
+            }
+        }
+        Some(target)
+    }
+
+    fn pointer_mut<'a>(&'a mut self, pointer: &str) -> Option<&'a mut Value> {
+        if pointer == "" {
+            return Some(self);
+        }
+        if !pointer.starts_with('/') {
+            return None;
+        }
+        let tokens = pointer.split('/').skip(1).map(|x| x.replace("~1", "/").replace("~0", "~"));
+        let mut target = self;
+
+        for token in tokens {
+            let tgt = target;
+            let target_opt = match *tgt {
+                Value::Object(ref mut map) => map.get_mut(&token[..]),
+                Value::Array(ref mut list) => {
+                    if let Some(idx) = parse_index(&token[..]) {
+                        list.get_mut(idx)
```

Comment:
Done.  I had originally had trouble with the closure, but it wasn't hard to correct (just needs a `move`) once I worked out the rest of the borrow checker's nagging. 😃